Add tutorial guide for annotated services

freevie opened the pull request:
Purpose of this patch:
To get feedback on the work on the tutorial guide from Armeria maintainers

Major changes:
- Tutorial guide itself
- Modifying <CodeBlock> component for highlights
- Supporting filename displays in code blocks with markdown syntax (```)

Please note:
- This work is still in progress.
- The location of the guides is subject to change.
- The commits contain Korean for notes and comments. They will be removed in near future.

Remaining work:
- Intro
- Guide in defining the data used in the sample service
- Links
- A section on finalizing the tutorial
- A section on tutorial summary
- Layout for referential information

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/line/armeria?pullRequest=3420) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/line/armeria?pullRequest=3420) before we can accept your contribution.<br/><hr/>**sooj-hwang** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/line/armeria?pullRequest=3420) it.</sub>

ikhoon commented:
Awesome! I'm glad to review this PR. 😊

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,134 @@
+    - Throw a <type://FallthroughException> at last, to try to convert other requests by the next converter.
+    - Feel free to add more mapping blocks in this request converter for other requests for your service.
+
+  <CodeBlock language="java" filename="BlogRequestConverter.java" highlight="9,10,13">{`
```
![image](https://user-images.githubusercontent.com/25103250/112917902-9e9cad00-913e-11eb-8bfb-643861773b7e.png)
Thanks a lot for implementing this highlighting feature!
I think there should be some spaces between the line number and the code.
I think @trustin might help you with this. 😄

trustin replied:
Haha, I could really use @ikhoon 's help :rofl:

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-intro.mdx:
```diff
@@ -0,0 +1,101 @@
+
+- JDK 14
+- Gradle
+  - Set your Gradle to compile Java with the `-parameters` option
```
I think JDK 8+ is enough. 😄 
The user does not have to preinstall the Gradle because the sample code provides [Gradle wrapper](https://docs.gradle.org/current/userguide/gradle_wrapper.html) that manages the Gradle build.
For the `-parameters`, let's link to it.
https://armeria.dev/docs/setup/#configure--parameters-javac-option

freevie replied:
Done!

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step1.mdx:
```diff
@@ -0,0 +1,86 @@
+
+## Next step
+
+Next, define a blog post object to use in requests and responses. (TW: 다음 페이지 표시를 어케 할까나..)
```
![image](https://user-images.githubusercontent.com/25103250/112918998-1f5ca880-9141-11eb-8cb8-1a47a6917c1a.png)

Probably the boxes at the bottom are enough?
So we can just leave the description of the next step as you have written.

freevie replied:
Thought so, too. Just in case readers find it hard to recognize the boxes at the bottom, I've left links in the content itself so they wouldn't have to wander for navigation.

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step2.mdx:
```diff
@@ -0,0 +1,60 @@
+
+TBU
+
+// TW Q: @RequestConverter(BlogPostRequestConverter.class)를 지금 추가해 주는 것이 좋은가?
```
Isn't it better to explain it in the `Implement CREATE operation` step with other request converters?

freevie replied:
Done :)

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step2.mdx:
````diff
@@ -0,0 +1,60 @@
+    }
+```
+
+// TW Q: 이 이해가 맞는가?
````
Yes, we can probably link to this page. https://www.baeldung.com/jackson-annotations#1-jsoncreator

freevie replied:
Link added

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,134 @@
+
+</Tip>
+
+// TW: 이 부분은 별도 섹션이 아닌 adomination으로 변경될 가능성 있음
```
What is adomination? 🤔

freevie replied:
Oh, this was a note for myself. Please ignore :)
BTW, sorry, it was a typo. Meant to type  admonition (e.g. warning, tip etc.)

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step6.mdx:
```diff
@@ -0,0 +1,160 @@
+  - HTTP method mapping (<type://@Delete>)
+  - Parameter injection (<type://@Param>)
+  - Exception handler (TBU)
+  - Blocking task executor (<type://@Blocking)
```
```suggestion
  - Blocking task executor (<type://@Blocking>)
```
Just in case you missed it. 😉

freevie replied:
Thank you, :) The list has been moved to somewhere else, but added the closing bracket.

minwoox reviewed (commented):
Thanks! I really love this tutorial. 😄

minwoox commented:
![screencapture-localhost-8000-tutorials-tutorial-blog-rest-intro-2021-03-30-10_15_36](https://user-images.githubusercontent.com/25103250/112920045-24baf280-9143-11eb-976e-02eda2926290.png)

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,169 @@
+4. Customize the default `convertRequest()` method as follows.
+    - Map the `title` and `content` properties of the body parameter to members in a `BlogPost` object, using the `stringValue()` method we wrote in the previous step.
+    - Throw the <type://FallthroughException> by calling <type://RequestConverterFunction#fallthrough()>, to pass on the request parameters that aren't converted by our converter to the default request converter.
+    - Feel free to add more mapping blocks (lines 9 to 11) in this request converter for other requests for the blog service.
```
Perhaps we don't need this sentence because it can confuse a user?

freevie replied:
Done https://github.com/line/armeria/pull/3420/commits/3949b2b3d73d89f5658908a153d1d94694ff43bd

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,169 @@
+
+### Assign a request converter
+
+In this step, assign the [request converter we customized](#write-a-request-converter) to our service method. Annotate the service method with <type://@RequestConverter> and specify the RequestConverterFunction class as `BlogPostRequestConverter.class`.
```
`<type://RequestConverterFunction>`

freevie replied:
Done https://github.com/line/armeria/pull/3420/commits/019ee257df96fc88302e69ba5b68b0552da13007

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,169 @@
+
+## 3. Implement service code
+
+When the request for creation is received, our request converter creates an instance of a blog post object for us. We want to save the blog post object in the map (`blogPosts`) created in the `BlogService` class. Skip adding a database operation for this tutorial.
```
I think we can remove the sentence `Skip adding ...`

freevie replied:
Done

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
````diff
@@ -0,0 +1,169 @@
+To store the blog post information in the map, add the following line in the `createBlogPost()` method.
+
+```java filename=BlogService.java highlight=2
+public HttpResponse createBlogPost(BlogPost blogPost) throws JsonProcessingException {
````
Let's add annotations(`@Post` and `@RequestConverter`) we have added in the previous example from now on.

freevie replied:
Done

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
````diff
@@ -0,0 +1,169 @@
+    ...
+  }
+  ```
+2. Create and return an HTTP response using Armeria's <type://HttpResponse> and include the information of the post created, as shown below, in the `createBlogPost()` method.
````
In my humble opinion, we can remove `as shown below` throughout the tutorials.

freevie replied:
All "as shown below"s are gone with the wind :)

minwoox reviewed (commented):
Thanks a lot, @freevie for working on this! 🙇 
I really like the overall composition of this tutorial.
Just left some suggestions and minor comments.

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-intro.mdx:
```diff
@@ -0,0 +1,110 @@
+
+Try writing the blog service yourself by following the instructions:
+
+- [Step 1. Create a server](./tutorial-blog-rest-step1)
```
Indeed, your suggestion is what our team guides for documentation.
Each guide is titled in gerunds, as you suggested (excluding navigation menus)
However, this particular list is a set of instructions, thus needed to be presented in imperatives. 
If we were to change the verbs to gerunds, then the introduction sentence should be changed to something like, "Try writing the blog service yourself with the following *guides*".

If this is the style for the Armeria site, then please let me know. I'll change the introduction sentence as well.

freevie replied:
Changed the the list as an ordered list + changed imperatives to gerunds.
Done 0f54ad46e495a1f26db334ed9c3ca8295b7cfcc7

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
````diff
@@ -0,0 +1,169 @@
+
+```java filename=BlogService.java highlight=1
+@Post("/blogs")
+public HttpResponse createBlogPost(BlogPost blogPost) throws JsonProcessingException {}
````
@minwoox 
I have a question. To keep the consistency with the rest of code samples on the site, 4-space indentations should be used. However, to save up space, I've used 2-space indentations on purpose. Please let me know your preference.

minwoox replied:
We use 4-space indentation in other examples and it's also our convention but I think it's okay to use 2-space indentation because:
- it can save up space as @freevie mentioned 
- there're people who use 2-space and we must not force our convention on them.
@trustin and @ikhoon any thought?

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,169 @@
+
+## What you need
+
+You need the BlogPost.java file, created from [step 2](./tutorial-blog-rest-step2). You can always [download](https://github.com/line/armeria-examples/blob/master/rest-api-annotated-service/src/main/java/example/armeria/server/blog/BlogPost.java) the full version of this file, instead of creating one yourself.
```
@minwoox ,
Question. Our team does not apply style on file names, based on the style guide which I can no longer remember the name of. But, I think I've seen somwhere on Armeria's site with styles applied on file names. Do you want me to apply the monospace style on all file names?

minwoox replied:
Let's do not use monospace for the file name.
Sorry about it. I was confused. 😄

freevie replied:
Got it. But, please do let me know if you wish to change'em all.

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,169 @@
+
+You need the BlogPost.java file, created from [step 2](./tutorial-blog-rest-step2). You can always [download](https://github.com/line/armeria-examples/blob/master/rest-api-annotated-service/src/main/java/example/armeria/server/blog/BlogPost.java) the full version of this file, instead of creating one yourself.
+
+## 0. Create a service file
```
This, I pondered upon for quite a while. All _implementing_ guides start with number 1, mapping with HTTP method. Starting here with 1 breaks consistency among implementing CRUD guides. However, starting here with 0 breaks consistency among all other guides. Let me think it over for a minute.

minwoox replied:
Ah then, either way is fine to me. 😄 
Step 1,2 and 7 does not start with `Map HTTP method` so I thought we do not have to align the the number one with `Map HTTP method`.

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step2.mdx:
```diff
@@ -0,0 +1,109 @@
+level: basic
+---
+
+# Preparing data files
```
Done https://github.com/line/armeria/pull/3420/commits/16869e847d7623ed901523f71f31b21e7835956b

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-intro.mdx:
```diff
@@ -0,0 +1,110 @@
+- [Step 2. Prepare data files](./tutorial-blog-rest-step2)
+- [Step 3. Implement CREATE](./tutorial-blog-rest-step3)
+- [Step 4. Implement READ](./tutorial-blog-rest-step4)
+- [Step 5. Implement PUT](./tutorial-blog-rest-step5)
```
Thank you for catching this. Must've been out of my mind.
Done https://github.com/line/armeria/pull/3420/commits/af44e88baf2aee7c3a64f407717148118a380856

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step1.mdx:
```diff
@@ -0,0 +1,107 @@
+package example.armeria.server.blog;
+
+import org.slf4j.Logger;            // for logger.info() calls
+import org.slf4j.LoggerFactory;     // for logger.info() calls
```
Done https://github.com/line/armeria/pull/3420/commits/291aa31c015fc4bd206905c2e1dc1754c4c560aa

ikhoon commented on site/src/pages/tutorials/tutorial-blog-rest-step2.mdx:
```diff
@@ -0,0 +1,109 @@
+
+In this step, we have implemented a data object specifying and containing blog post objects.
+
+Next, at [Step 3. Implement CREATE](./tutorial-blog-rest-step3), you'll implement a CREATE operation to create blog posts.
```
Could we globally use paths that start with `/tutorials/tutorial-blog-rest-step3` instead of `../../path-to-page` style?

freevie replied:
Done! b23f747c9e5c8d2e3a2f01116ce3fe9a48f2a701
(Although, I haven't checked if all links work fine. Will make a final check on it)

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step1.mdx:
```diff
@@ -0,0 +1,107 @@
+Start by creating a main class:
+
+1. Create a Java file named Main. For the full version of the file, check the [service sample code](https://github.com/line/armeria-examples/blob/master/rest-api-annotated-service/src/main/java/example/armeria/server/blog/Main.java).
+2. Add required import statements as shown below.
```
Done https://github.com/line/armeria/pull/3420/commits/88f43949f1ab18dbeb56a9a444a9c52c87f74a20

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step1.mdx:
```diff
@@ -0,0 +1,107 @@
+
+## 1. Create the main class
+
+Start by creating a main class:
```
Actually, I went through the tutorial and removed most of them before this commit, but I think it may not be too bad to add them back. However, I avoided repeating them too much in each step.
Plus, I've changed bits and pieces to make the tone lighter than before.
Done e330b463100dd63e5f1e299ceb0e374ba68f35a1

minwoox replied:
Thanks for the tone! 😄 Let me take a look at them soon. ❤️

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step1.mdx:
```diff
@@ -0,0 +1,107 @@
+
+## 2. Create a server instance
+
+To create a server instance, use Armeria's <type://ServerBuilder>. Add the method, `newServer()`, as shown below, in the `Main` class. Since server instances require at least one service to run with, add a dummy service returning `"Hello, Armeria!"` for the mere purpose of checking that the server gets launched and runs.
```
Done 38f4dda5c26eaee05a0e2200e7e5659280001c33

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step1.mdx:
```diff
@@ -0,0 +1,107 @@
+  return sb.http(port)
+            .service("/",(ctx, req) ->
+                HttpResponse.of("Hello, Armeria!"))
+            .build();
```
Done! Thank you for the tip
ce56d0ac01999ebd741a7f6ccc18b19ed2d9544b

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,169 @@
+We can use Armeria's default request converter as is, but here let's give a go at customizing a request converter for our blog post requests. We want to convert blog post details into a Java object.
+
+1. Create a BlogPostRequestConverter.java file.
+2. Declare a class in the file, implementing the <type://RequestConverterFunction> interface. For simplicity sake, generate impromptu IDs in this tutorial.
```
Done 23ac3c171c422db63351f084a5af31340da3a56b

freevie commented on site/src/components/steps.module.less:
```diff
@@ -0,0 +1,20 @@
+
+  .ant-steps-horizontal {
```
Done 00b4536168fb924e0c40ca40bc55f90614cfd941

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step1.mdx:
```diff
@@ -0,0 +1,107 @@
+
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.server.Server;
+import com.linecorp.armeria.server.ServerBuilder;
```
Ah, we don't use the doc service in this step, but this reminds me to add it for step 7. (Adding services to server)
I'll also remove the following from the code snippet from the step1 file.
```
    final DocService docService = DocService.builder()
                                            .build();
```

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step1.mdx:
```diff
@@ -0,0 +1,107 @@
+                                          .build();
+  return sb.http(port)
+            .service("/",(ctx, req) ->
+                HttpResponse.of("Hello, Armeria!"))
```
Minwoo has also given the same feedback, so this has been applied (25a88e63671179c598ec5b3e58fc7177c98a83c0)

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step2.mdx:
```diff
@@ -0,0 +1,109 @@
+                      .add("createdAt", createdAt)
+                      .add("modifiedAt", modifiedAt)
+                      .toString();
+}
```
Removed! 6e63b49ef0f66c2e32afa506b23eb2c2ac675e48

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step1.mdx:
```diff
@@ -0,0 +1,107 @@
+static Server newServer(int port) {
+  final ServerBuilder sb = Server.builder();
+  final DocService docService = DocService.builder()
+                                          .build();
```
I left this by mistake. Thank you for catching. As I left a comment earlier, we won't be adding the DocService at this stage, so I 'd removed it from the code. (3c68fb309056a3a5813737088fac320f6887010f)

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,169 @@
+
+### Write a request converter
+
+Armeria's [request converter](../docs/server-annotated-service#converting-an-http-request-to-a-java-object) converts a request body from a client into a Java object for you.
```
Applied! 583c7b176c4d00db845fb515c876091e50c5ad2c

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step2.mdx:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  BlogPost(int id, String title, String content, long createdAt) {
+    this(id, title, content, createdAt, createdAt);
```
Removed the sections dividing the two. However, a description separates the two code snippets.
Please let me know if you wish to have the snippet merged. --> 4d08ebab294af156118ed13110d62413286f20db

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,169 @@
+## 0. Create a service file
+
+1. Create a service file, BlogService.java.
+2. Declare a class, `BlogService`, in the file as shown below. Import required files as you go along the tutorial or copy all from the [sample file](https://github.com/line/armeria-examples/blob/master/rest-api-annotated-service/src/main/java/example/armeria/server/blog/BlogService.java).
```
Removed the steps, and merged the two instructions into one. However, removed the redundancy by leaving the filename only, as you suggested.
4e803e0a0e5b75937fbd9b8f8bc7be33e834ef07

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
````diff
@@ -0,0 +1,169 @@
+2. Declare a class, `BlogService`, in the file as shown below. Import required files as you go along the tutorial or copy all from the [sample file](https://github.com/line/armeria-examples/blob/master/rest-api-annotated-service/src/main/java/example/armeria/server/blog/BlogService.java).
+  ```java filename=BlogService.java
+  public final class BlogService {
+
````
I take it that you want me to add this comment for the object mapper line, which I did.
85d8f4b59a4289e8c814eaca8ada308d697759c8
Please let me know if I got you wrong.

minwoox replied:
Yes, that's exactly what I wish to do. Sorry about the unclear comment. 😓

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,169 @@
+
+## 1. Map HTTP method
+
+Start by [mapping HTTP POST method](/docs/server-annotated-service#mapping-http-service-methods) with our service method, which we will declare at this step:
```
Left the links in the introduction line, but removed them from step instructions - cd44f151846232143f901b7f4a1f0c37726b49a1

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step4.mdx:
```diff
@@ -0,0 +1,132 @@
+1. [Create a service file](/tutorials/tutorial-blog-rest-step3#0-create-a-service-file) if you're yet to create one.
+2. Declare service methods, `getBlogPost()` and `getBlogPosts()` in the class `BlogService`.
+3. Map these service methods with the HTTP GET method by adding the <type://@Get> annotation as follows.
+4. Assign the endpoint as `/blogs`.
```
nit:
Is `bind`  more appropriate for this context?
`Bind the endpoint to /blogs.` 🤔

freevie replied:
Changed the pattern to "Bind the endpoing `/abc` to the method. Hope this is okay.
bdcd5ad30585e3939c4effbc35616de7cc7501fe

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step6.mdx:
```diff
@@ -0,0 +1,154 @@
+
+### Add blocking
+
+With real services, accessing and operating on a database takes time. To prevent the service to wait on our task, we want to inform the task executor to handle other tasks until we're done with database operation.
```
How about
```suggestion
With real services, accessing and operating on a database takes time. We need to hand over those blocking tasks to [blocking task executor](/docs/server-annotated-service#specifying-a-blocking-task-executor) so that the EventLoop isn't blocked.
```

freevie replied:
Applied the suggestion, except that I've changed "those" to "such".
Done 264622773062b7bbac0f56eab5f5505fb9e376e9

minwoox replied:
> I've changed "those" to "such".

That's way better. 😄

minwoox commented:
> Haha, I could really use @ikhoon 's help 🤣

@freevie has done it by herself. Thanks a lot!
If it was me, it might take a day to fix that. 🤣

minwoox reviewed (commented):
Left some minor comments. I think it's getting perfect. 👍

ikhoon commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,168 @@
+
+We can use Armeria's default request converter as is, but here let's give a go at customizing a request converter for our blog post requests. We want to convert blog post details into a Java object.
+
+1. Create a BlogPostRequestConverter.java file and declare a class, implementing the <type://RequestConverterFunction> interface. For simplicity sake, generate impromptu IDs for this tutorial.
```
```suggestion
1. Create a BlogPostRequestConverter.java file and declare a class, implementing the <type://RequestConverterFunction> interface. For simplicity's sake, generate impromptu IDs for this tutorial.
```

freevie replied:
Simplicity's sake doesn't sound grammatically correct to me, somehow.
Changed it to "For the sake of simplicity". Hope this is alright. -> c917dd347bbbe0469e0549a35fbeb3a9bba4df60

ikhoon commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
````diff
@@ -0,0 +1,168 @@
+  ```java filename=BlogPost.java highlight=3,6,9,12,15
+  final class BlogPost {
+    ...
+    @JsonProperty("id")
+    int id() { return id; }
+
+    @JsonProperty("title")
+    String title() { return title; }
+
+    @JsonProperty("content")
+    String content() { return content; }
+
+    @JsonProperty("createdAt")
+    long createdAt() { return createdAt; }
+
+    @JsonProperty("modifiedAt")
+    long modifiedAt() { return modifiedAt; }
````
Actually, these are not a Java standard getter style.
How about using `getXXX()` style? If so, you can also remove `@JsonProperty("XXX")` for simplicity's sake.

freevie replied:
Ah, I believe I'll need @minwoox 's comment on this, since the code samples in the tutorial will have to be identical to the service sample provided on Git.

minwoox replied:
Yeah, this is Armeria style. Let's use `getXXX()` style and remove `@JsonProperty("XXX")` as well.
I will make changes to the sample code. Sorry about it. 😅

freevie replied:
@minwoox , @ikhoon 
In such case, I get an error for not having a serializer. Should I add `@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)` for the `BlogPost` class or use `@JsonSetter` for each `getOOO()`? Or is there any other alternatives I should use? (Sorry, I am not familiar with this area)

minwoox replied:
Oops let me see what happens

minwoox replied:
Ah we have to make it public. 😅 
```
public final class BlogPost {
  ...
  public int getId() {...}

  public String getTitle() {...}
  ...
```

freevie replied:
Works like a charm! Thank you
Done 8d694d264d77c7bded67373d06cd348835ff38f1

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step4.mdx:
````diff
@@ -0,0 +1,132 @@
+```java filename=BlogService.java highlight=3-4
+public HttpResponse getBlogPost(@Param int id) throws JsonProcessingException {
+  ...
+  final byte[] content = mapper.writeValueAsBytes(blogPost);
````
Aye aye captain! Hope I haven't left any `final`s behind.
d9b45e038a5b541710233c1d9b142fabb26c98ba
beff9bf7acd1ca0c611fde6a718df51644bb87fb (Restored the keyword for ObjectMapper)

freevie replied:
Thanks for catching!
Here is the missing `final` for the `logger` -> d94be1d5d2c108bf23cd097fb1f42f041fc85505

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-intro.mdx:
```diff
@@ -0,0 +1,110 @@
+| Create | `createBlogPost()` | `@Post` |
+| Read | `getBlogPost()`, `getBlogPosts()` | `@Get` |
+| Update | `updateBlogPost()` | `@Put` |
+| Delete | `deleteBlogPost()` | `@Delete` |
```
Done! bfabedfc8ed021abcf1defc4e5188da4c81d066e

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-intro.mdx:
```diff
@@ -0,0 +1,110 @@
+## Run sample service
+
+Have a go at running the sample service and experience the outcome of this tutorial.
+Using Armeria's Doc service, you can see a server running, receiving requests and providing responses.
```
Wonderful suggestion! Thank you
Done cdde00c8f5ac50e4c55ea69ca981a9d3d2a37c65

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-intro.mdx:
````diff
@@ -0,0 +1,110 @@
+  ```bash
+  $ ./gradlew run
+  ```
+4. Open the Doc service on your web browser at http://127.0.0.1:8080/docs.
````
Added `<type://DocService>` here for variation.
Done cdde00c8f5ac50e4c55ea69ca981a9d3d2a37c65

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,168 @@
+
+Let's take post information through a request body. Armeria's [request converter](/docs/server-annotated-service#converting-an-http-request-to-a-java-object) converts request parameters in HTTP messages into Java objects for you. In the request converter we define what keys of a JSON object to map with what properties of a Java object.
+
+To handle parameters received for creating a blog post, use <type://RequestConvert>, in these steps. The order may be changed.
```
Thank you!
Done e2b73af3d9d72f082c5a884e0a860ea7612d1e4d

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,168 @@
+
+Now, it's time to return a response to our client. As the response, on top of the information received, add the ID of the post, created time, plus modified time which would be identical to the created time.
+
+Let's a return a response for blog post creation:
```
My bad!
Done 4efdf6a226e6f54797571ee422afbcd34b365583

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step4.mdx:
````diff
@@ -0,0 +1,132 @@
+To retrieve multiple blog posts, copy the following code inside the `getBlogPosts()` method:
+
+```java filename=BlogService.java highlight=3-10
+public Iterable<BlogPost> getBlogPosts(@Param @Default("true") boolean descending) {
````
@minwoox 
Sorry, I just wanted to confirm if we are guiding readers to add the configuration you mentioned in the `build.gradle` file?

freevie replied:
For now, added the import statement for getting multiple blog posts -> 5b4d87be6dc498e9d06844c556578a0709db0ad3

Will add the Gradle explanation soon and leave a comment here soon.

minwoox replied:
> Sorry, I just wanted to confirm if we are guiding readers to add the configuration you mentioned in the build.gradle file

Yes and sorry about not adding the dependencies to the build.gradle of the example.
It's automatically added so I missed adding them explicitly. 😅

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step6.mdx:
````diff
@@ -0,0 +1,152 @@
+To delete a blog post, copy the code below to the `deleteBlogPost()` method.
+
+```java filename=BlogService.Java highlight=2
+public HttpResponse deleteBlogPost(@Param int id) {
````
ditto. Let's keep `@Delete("/blogs/:id")`

freevie replied:
Had been added along with adding import statements, 48b7d95931408b1e9f0d0fb877597525ffb443f8

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step6.mdx:
````diff
@@ -0,0 +1,152 @@
+  }
+  ```
+2. Create an exception handler for the blog service:
+  1. Create a file, BadRequestExceptionHandler.java
````
![image](https://user-images.githubusercontent.com/25103250/122151581-c62c2780-ce9a-11eb-9198-1293f86e826f.png)
This does not render properly. We need an extra space. 
```suggestion
   1. Create a file, BadRequestExceptionHandler.java
```

freevie replied:
Had been added along with adding import statements, 48b7d95931408b1e9f0d0fb877597525ffb443f8

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step6.mdx:
````diff
@@ -0,0 +1,152 @@
+  ```java filename=BlogService.java highlight=1
+  @Delete("/blogs/:id")
+  @ExceptionHandler(BadRequestExceptionHandler.class)
+  public HttpResponse deleteBlogPost(@Param int id) {
+    ...
````
```suggestion
  public HttpResponse deleteBlogPost(@Param int id) {
    ...
  }
```

minwoox replied:
or
```suggestion
  public HttpResponse deleteBlogPost(@Param int id) { ... }
```

freevie replied:
Opted the second suggestion, thank you.
34b2de0ac1c4ddf269c992308bf94c30353de7ec

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step6.mdx:
```diff
@@ -0,0 +1,152 @@
+@Blocking
+@Delete("/blogs/:id")
+@ExceptionHandler(BadRequestExceptionHandler.class)
+public HttpResponse deleteBlogPost(@Param int id) {
```
```suggestion
public HttpResponse deleteBlogPost(@Param int id) { ...}
```

freevie replied:
Done 2b82f772c834552d110766b76f22106dba273ccd

minwoox reviewed (commented):
Thanks! Left some minor comments and suggestions. 🙇

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-intro.mdx:
```diff
@@ -0,0 +1,132 @@
+
+### build.gradle
+
+The sample service uses third party libraries, Guava and Jackson. They are added as dependencies in the build.gradle file.
```
Done 6226d0abf6af01dc59a6f34d3451565f95740430

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-intro.mdx:
```diff
@@ -0,0 +1,132 @@
+## Run sample service
+
+Have a go at running the sample service and experience the outcome of this tutorial.
+Using Armeria's [Documentation Service](/docs/server-docservice), you can see a server running, receiving requests and providing responses.
```
Done 5c88ae43f17b9fd52aae7e5174f8ea4ee8642529

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step1.mdx:
```diff
@@ -0,0 +1,98 @@
+
+Let's create a server instance using Armeria's <type://ServerBuilder>:
+
+1. Create a main class and add import statements. You can see the full version of the file [here](https://github.com/line/armeria-examples/blob/master/tutorials/rest-api-annotated-service/src/main/java/example/armeria/server/blog/Main.java).
```
Done 44a8903169f1ab9c34015d5bcef7ad8b47d4a981

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step2.mdx:
````diff
@@ -0,0 +1,88 @@
+Let's define a `BlogPost` class containing the following information.
+
+```java
+final class BlogPost {
````
Done e1ae0ee01798500a00e8f4da55d642b9048f8f8d

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step2.mdx:
````diff
@@ -0,0 +1,88 @@
+Add the following block of code in the blog service.
+
+```java
+final class BlogPost {
````
Done e1ae0ee01798500a00e8f4da55d642b9048f8f8d

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step2.mdx:
````diff
@@ -0,0 +1,88 @@
+}
+```
+
+## 3. Add methods
````
Done b3668e7628a4f0112969ce49cccf0b809fc33c56

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,148 @@
+
+Let's receive blog post information through a request body. Armeria's [request converter](/docs/server-annotated-service#converting-an-http-request-to-a-java-object) converts request parameters in HTTP messages into Java objects for you. In the request converter we define what keys of a JSON object to map with what properties of a Java object.
+
+Let's first [write a request converter](#write-a-request-converter) and then [register the request converter](#register-a-request-converter) the service method.
```
Done 1112632c708b2623cd6b2d536d4d653f9d0a0e46

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
````diff
@@ -0,0 +1,148 @@
+  }
+  ```
+
+## Next step
````
@minwoox 
All the testing has been put to the step 7, using the `DocService`.
But I take it that you want to add a section with code samples.

Will it be okay to reuse the code in the BlogServiceTest.java?

minwoox replied:
code, curl, and `DocService` are all fine to me. 😄

minwoox replied:
> All the testing has been put to the step 7, using the DocService.

We can test all in once in step 7, but I thought a user wants to check if he/she implements the method correctly or not at the end of each step.
So if we provide a separate section using code, curl or `DocService`, a user can easily follow that. 🤔

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step4.mdx:
````diff
@@ -0,0 +1,138 @@
+}
+```
+
+We've already implemented returning multiple blog posts in the previous step. As a final touch, annotate the method `getBlogPosts()` with <type://@ProducesJson>. This returns the response as a JSON object for you.
````
Changed to:

> This converts a list of `BlogPost` objects into a JSON response.

05dd35211d5912e35e720317012027e217843cb3

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step5.mdx:
````diff
@@ -0,0 +1,128 @@
+2. [Inject the path parameter](/docs/server-annotated-service#parameter-injection) to the service method by annotating the parameter with <type://@Param>.
+3. [Convert request body](/docs/server-annotated-service#converting-an-http-request-to-a-java-object) into a Java object by annotating the `BlogPost` parameter with <type://@RequestObject>.
+  ```java filename=BlogService.java
+  // Instructions 1 to 3
````
@minwoox 
Removing them will be no problem, but then the code for this section has annotations that weren't there in the previous mention (i.e. Mapping HTTP method).

- Is it okay with you?
- If so, did you want to leave `@Param` and `@RequestObject` in "2. Handle parameters" or move them to "1. Map HTTP method"?

minwoox replied:
What I meant was removing `// Instructions 1 to 3`. 😄

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step5.mdx:
````diff
@@ -0,0 +1,128 @@
+  ```java filename=BlogPost.java
+  @JsonCreator
+  BlogPost(@JsonProperty("id") int id, @JsonProperty("title") String title,
+            @JsonProperty("content") String content) {
````
Done 47d5d5d6ee1f584475adb1eab7dcee691fab4311

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step5.mdx:
````diff
@@ -0,0 +1,128 @@
+To update a blog post, copy the following code inside the `updateBlogPost()` method.
+
+```java filename=BlogService.java
+  public HttpResponse updateBlogPost(@Param int id, @RequestObject BlogPost blogPost){
````
Had been added along with adding import statements, 48b7d95931408b1e9f0d0fb877597525ffb443f8

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step5.mdx:
````diff
@@ -0,0 +1,128 @@
+Inside the condition block of the `getBlogPost()` method added in the previous step, return a response using Armeria's <type://HTTPResponse>, with the content of the response as `HttpStatus.NOT_FOUND`.
+
+```java filename=BlogService.java highlight=4
+public HttpResponse updateBlogPost(@Param int id, @RequestObject BlogPost blogPost){
````
Done 3cee64fd97c70d33c63d1815f6a91e82e2504f1e

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step5.mdx:
````diff
@@ -0,0 +1,128 @@
+Let's return the information updated when the blog post exists. To return the updated information, copy the code below to the `updateBlogPost()` method.
+
+```java filename=BlogService.java
+public HttpResponse updateBlogPost(@Param int id, @RequestObject BlogPost blogPost){
````
Had been added along with adding import statements, 48b7d95931408b1e9f0d0fb877597525ffb443f8

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-step4.mdx:
````diff
@@ -0,0 +1,138 @@
+}
+```
+
+Now, let's retrieve multiple blog posts:
````
Separate the content into two tabs c8e0945ccac88eaebe256034e1bf89ea616b651a

+ Am working on sending a sample request for steps 3-6 :)

freevie replied:
Changed the tab separation structure https://github.com/line/armeria/pull/3420/commits/49e54d306699ac1ff1b36c3a294cf3cdc6f77151

freevie commented:
@minwoox ,

I've added testing parts in steps 3-6, 49e54d306699ac1ff1b36c3a294cf3cdc6f77151, with changes here and there.

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,223 @@
+import java.util.concurrent.ConcurrentHashMap;
+
+public final class BlogService {
+  private Map<Integer, BlogPost> blogPosts = new ConcurrentHashMap<>();
```
```suggestion
  private final Map<Integer, BlogPost> blogPosts = new ConcurrentHashMap<>();
```

freevie replied:
Done 17d166458b83a88a5823fa99909280ab1fed1ae9

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,223 @@
+  ...
+
+  @Post("/blogs")
+  public void createBlogPost(BlogPost blogPost)
+      throws JsonProcessingException {}
```
```suggestion
  public void createBlogPost(BlogPost blogPost) {}
```

freevie replied:
Done 17d166458b83a88a5823fa99909280ab1fed1ae9

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
````diff
@@ -0,0 +1,223 @@
+```java filename=BlogService.java highlight=2
+@Post("/blogs")
+@RequestConverter(BlogPostRequestConverter.class)
+public void createBlogPost(BlogPost blogPost)
+    throws JsonProcessingException {
````
```suggestion
public void createBlogPost(BlogPost blogPost) {
```

freevie replied:
Done 17d166458b83a88a5823fa99909280ab1fed1ae9

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
````diff
@@ -0,0 +1,223 @@
+```java filename=BlogService.java highlight=5 showlineno=true
+@Post("/blogs")
+@RequestConverter(BlogPostRequestConverter.class)
+public void createBlogPost(BlogPost blogPost)
+    throws JsonProcessingException {
````
```suggestion
public void createBlogPost(BlogPost blogPost) {
```

freevie replied:
Done 17d166458b83a88a5823fa99909280ab1fed1ae9

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
```diff
@@ -0,0 +1,223 @@
+  import com.linecorp.armeria.common.HttpResponse;
+  import com.linecorp.armeria.common.HttpStatus;
+
+  public HttpResponse createBlogPost(BlogPost blogPost)
+      throws JsonProcessingException {
```
```suggestion
  public HttpResponse createBlogPost(BlogPost blogPost) {
```

freevie replied:
Done 17d166458b83a88a5823fa99909280ab1fed1ae9

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
````diff
@@ -0,0 +1,223 @@
+3. Bind the endpoint `/blogs` to the method.
+
+```java filename=BlogService.java highlight=7
+import com.fasterxml.jackson.core.JsonProcessingException;
````
```suggestion
```

freevie replied:
Done 17d166458b83a88a5823fa99909280ab1fed1ae9

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
````diff
@@ -0,0 +1,223 @@
+
+  ```java filename=BlogService.java highlight=4,7
+  import com.linecorp.armeria.common.HttpResponse;
+  import com.linecorp.armeria.common.HttpStatus;
````
```suggestion
```

freevie replied:
Done 187468b350174ff365ddb38fdf3fb73f2a3e0fe6

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
````diff
@@ -0,0 +1,223 @@
+3. Call the service method for creating a blog post. Here, we'll use the cURL command.
+
+  ```bash
+  $ curl --location --request POST 'localhost:8080/blogs' \
````
```suggestion
  $ curl --request POST 'localhost:8080/blogs' \
```
I think we don't have to consider redirect now. 😄

freevie replied:
Done ab83cf2e802577fed3da3f8085929d464bbeb313

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step3.mdx:
````diff
@@ -0,0 +1,223 @@
+4. Check the return value.
+
+  ```bash
+  {"id":0,"title":"My first blog","content":"Hello Armeria!","createdAt":1624489646371,"modifiedAt":1624489646371}
````
nit: let's just use `...` for all timestamps. 😉 
```suggestion
  {"id":0,"title":"My first blog","content":"Hello Armeria!","createdAt":...,"modifiedAt":...}
```

freevie replied:
Done ab83cf2e802577fed3da3f8085929d464bbeb313

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step5.mdx:
```diff
@@ -0,0 +1,259 @@
+
+<Warning>
+
+To test retrieving a blog post, you need to have blog posts created, which requires you to have the creation method implemented in the BlogService.java. See [Step 3. Implement CREATE](/tutorials/tutorial-blog-rest-step3) for instructions.
```
```suggestion
To test updating a blog post, you need to have blog posts created, which requires you to have the creation method implemented in the BlogService.java. See [Step 3. Implement CREATE](/tutorials/tutorial-blog-rest-step3) for instructions.
```

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step5.mdx:
````diff
@@ -0,0 +1,259 @@
+3. Bind the endpoint `/blogs` to the method.
+
+```java filename=BlogService.java highlight=7
+import com.fasterxml.jackson.core.JsonProcessingException;
````
```suggestion
```

freevie replied:
Done 90f0b0f382f94a474bfd0d0ccb63bdb362d3dd6c

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step5.mdx:
```diff
@@ -0,0 +1,259 @@
+  ...
+
+  @Put("/blogs")
+  public void updateBlogPost(int id, BlogPost blogPost)
+          throws JsonProcessingException {
```
```suggestion
  public void updateBlogPost(int id, BlogPost blogPost) {
```

freevie replied:
Done 90f0b0f382f94a474bfd0d0ccb63bdb362d3dd6c

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step5.mdx:
````diff
@@ -0,0 +1,259 @@
+  ```java filename=BlogService.java
+  // Instructions 1 to 3
+  @Put("/blogs/:id")
+  public void updateBlogPost(@Param int id, @RequestObject BlogPost blogPost)
+      throws JsonProcessingException {
````
```suggestion
  public void updateBlogPost(@Param int id, @RequestObject BlogPost blogPost) {
```

freevie replied:
Done 90f0b0f382f94a474bfd0d0ccb63bdb362d3dd6c

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step5.mdx:
```diff
@@ -0,0 +1,259 @@
+  import com.fasterxml.jackson.annotation.JsonCreator;
+  import com.fasterxml.jackson.annotation.JsonProperty;
+
+  final class BlogPost {
```
```suggestion
  public final class BlogPost {
```

freevie replied:
Done 90f0b0f382f94a474bfd0d0ccb63bdb362d3dd6c

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step5.mdx:
````diff
@@ -0,0 +1,259 @@
+
+```java filename=BlogService.java
+  @Put("/blogs/:id")
+  public void updateBlogPost(@Param int id, @RequestObject BlogPost blogPost){
````
```suggestion
  public void updateBlogPost(@Param int id, @RequestObject BlogPost blogPost) {
```

freevie replied:
Done 90f0b0f382f94a474bfd0d0ccb63bdb362d3dd6c

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step5.mdx:
```diff
@@ -0,0 +1,259 @@
+Let's return an error for the request to update a blog post that doesn't exist.
+
+1. Replace the return type of the `updateBlogPost()` method from `void` to `HttpResponse`.
+2. Return a response using Armeria's <type://HTTPResponse> containing `HttpStatus.NOT_FOUND`.
```
```suggestion
2. Return a response using Armeria's <type://HttpResponse> containing `HttpStatus.NOT_FOUND`.
```

freevie replied:
Done 90f0b0f382f94a474bfd0d0ccb63bdb362d3dd6c

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step5.mdx:
```diff
@@ -0,0 +1,259 @@
+public final class BlogService {
+  ...
+  @Put("/blogs/:id")
+  public HttpResponse updateBlogPost(@Param int id, @RequestObject BlogPost blogPost){
```
```suggestion
  public HttpResponse updateBlogPost(@Param int id, @RequestObject BlogPost blogPost) {
```

freevie replied:
Done 90f0b0f382f94a474bfd0d0ccb63bdb362d3dd6c

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step5.mdx:
````diff
@@ -0,0 +1,259 @@
+
+```java filename=BlogService.java highlight=8,10
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
````
```suggestion
```

freevie replied:
Done 90f0b0f382f94a474bfd0d0ccb63bdb362d3dd6c

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step5.mdx:
```diff
@@ -0,0 +1,259 @@
+  ...
+
+  @Put("/blogs/:id")
+  public HttpResponse updateBlogPost(@Param int id, @RequestObject BlogPost blogPost){
```
```suggestion
  public HttpResponse updateBlogPost(@Param int id, @RequestObject BlogPost blogPost) {
```

freevie replied:
Done 90f0b0f382f94a474bfd0d0ccb63bdb362d3dd6c

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-step6.mdx:
```diff
@@ -0,0 +1,277 @@
+
+<Warning>
+
+To test retrieving a blog post, you need to have blog posts created, which requires you to have the creation method implemented in the BlogService.java. See [Step 3. Implement CREATE](/tutorials/tutorial-blog-rest-step3) for instructions.
```
```suggestion
To test deleting a blog post, you need to have blog posts created, which requires you to have the creation method implemented in the BlogService.java. See [Step 3. Implement CREATE](/tutorials/tutorial-blog-rest-step3) for instructions.
```

minwoox reviewed (commented):
Thanks! left some minor comments.
I think it's really close to the end. 😄

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-intro.mdx:
```diff
@@ -0,0 +1,132 @@
+}
+
+dependencies {
+  implementation "com.linecorp.armeria:armeria:${versions['com.linecorp.armeria:armeria-bom']}"
```
Done 64c1ac504500dbe0a06ab1ecd15462c12ce7b34d

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-intro.mdx:
```diff
@@ -0,0 +1,132 @@
+This tutorial is based on a [sample service](https://github.com/line/armeria-examples/tree/master/tutorials/rest-api-annotated-service), a minimal blog service, with which you can create, read, update, and delete blog posts.
+To keep our focus on Armeria, this tutorial and the sample service use memory-based operations instead of database operations.
+
+[Try writing a service yourself](#try-writing-blog-service-yourself) by following this tutorial or have a go at [running the sample service](#run-sample-service) right away.
```
Done 81ebbfbcfa491f8f0a7341ede9dc91a6307be479

freevie commented:
@minwoox 

I have removed the instruction for adding the service to the server in "Test" sections from each CRUD operation guide. However, to follow your suggestion to _additionally_ adding the blog service in the 1. Creating a server section, we needed to have the service file (BlogService.java) created also in the Step 1 guide, which seemed a bit complicated. 

So, long story short, I have removed the old Step 7 (Adding services to the server) up to Step 3. Which enabled removing the instruction for creating a service file (that was indexed with 0) from each CRUD implementation guide.
I'll give you the details in a chat tomorrow.

minwoox commented on site/src/pages/tutorials/tutorial-blog-rest-update.mdx:
````diff
@@ -0,0 +1,229 @@
+  ```bash
+  $ curl --request POST 'localhost:8080/blogs' \
+  -H 'Content-Type: application/json' \
+  -d '{"title":"Test blog for retrieving", "content":"Create to read."}'
````
nit: updating and update?

freevie replied:
Changed the whole example text for READ, UPDATE, and DELETE (62784099df81ece9305254ec75ed62bdc35e480f)

minwoox reviewed (approved):
🎉 🎉 🎉 
Thanks a lot for your hard work, @freevie!

ikhoon commented:
@freevie Could you run `npm run lint:fix` to fix some lint violations?

ikhoon commented on site/src/pages/tutorials/tutorial-blog-rest-create.mdx:
```diff
@@ -0,0 +1,192 @@
+
+## 2. Handle parameters
+
+Let's receive blog post information through a request body. Armeria's [request converter]
+(/docs/server-annotated-service#converting-an-http-request-to-a-java-object) converts request parameters in HTTP messages into Java objects for you. In the request converter, we define what keys of a JSON object to map with what properties of a Java object.
```
An href link for the following link will not be generated cause there is a new line between `[..]` and `(...)`.
```suggestion
Let's receive blog post information through a request body. Armeria's [request converter](/docs/server-annotated-service#converting-an-http-request-to-a-java-object)
converts request parameters in HTTP messages into Java objects for you. In the request converter, we define what keys of a JSON object to map with what properties of a Java object.
```

ikhoon commented on site/src/pages/tutorials/tutorial-blog-rest-create.mdx:
```diff
@@ -0,0 +1,192 @@
+
+Armeria's [request converter](/docs/server-annotated-service#converting-an-http-request-to-a-java-object) converts a request body from a client into a Java object for you.
+
+We can use Armeria's default request converter as is, but here let's give a go at customizing a request converter for our blog post requests. We want to convert blog post details into a Java object.
```
How about?
```suggestion
We can use Armeria's default <type://JacksonRequestConverterFunction> as is, but here let's give a go at customizing a request converter for our blog post requests. We want to convert blog post details into a Java object.
```

ikhoon commented on site/src/pages/tutorials/tutorial-blog-rest-create.mdx:
```diff
@@ -0,0 +1,192 @@
+
+  final class BlogPostRequestConverter implements RequestConverterFunction {
+    ...
+    @Override
+
+    public Object convertRequest(ServiceRequestContext ctx,
```
```suggestion
    @Override
    public Object convertRequest(ServiceRequestContext ctx,
```

ikhoon commented on site/src/pages/tutorials/tutorial-blog-rest-read.mdx:
```diff
@@ -0,0 +1,275 @@
+  ...
+
+  @Get("/blogs")
+  public void getBlogPost(int id)
+      throws JsonProcessingException {
```
```suggestion
  public void getBlogPost(int id) {
```

ikhoon commented on site/src/pages/tutorials/tutorial-blog-rest-read.mdx:
````diff
@@ -0,0 +1,275 @@
+
+```java filename=BlogService.java showlineno=true
+@Get("/blogs/:id")
+public void getBlogPost(@Param int id)
+    throws JsonProcessingException {
````
```suggestion
public void getBlogPost(@Param int id) {
```

ikhoon commented on site/src/pages/tutorials/tutorial-blog-rest-read.mdx:
````diff
@@ -0,0 +1,275 @@
+
+```java filename=BlogService.java highlight=2,6
+@Get("/blogs/:id")
+public HttpResponse getBlogPost(@Param int id)
+    throws JsonProcessingException {
````
```suggestion
public HttpResponse getBlogPost(@Param int id) {
```

ikhoon commented:
A horizontal scroll seems not working. 
<img width="800" alt="image" src="https://user-images.githubusercontent.com/1866157/124559014-30226600-de76-11eb-8ef2-3589382b4827.png">

ikhoon commented on site/src/pages/tutorials/tutorial-blog-rest-delete.mdx:
```diff
@@ -0,0 +1,241 @@
+We've already handled returning the not found error in the [exception handling section](#handle-exceptions). Here, we'll return a response for successful deletion.
+
+1. Replace the return type of the `deleteBlogPost()` method from `void` to `HttpResponse`.
+2. Return a response using Armeria's <type://HttpResponse>, containing `HttpStatus.NO_CONTENT`.
```
```suggestion
2. Return a response using Armeria's <type://HttpResponse>, containing <type://HttpStatus#NO_CONTENT>.
```

ikhoon commented on site/src/pages/tutorials/tutorial-blog-rest-delete.mdx:
```diff
@@ -0,0 +1,241 @@
+
+We've already handled returning the not found error in the [exception handling section](#handle-exceptions). Here, we'll return a response for successful deletion.
+
+1. Replace the return type of the `deleteBlogPost()` method from `void` to `HttpResponse`.
```
```suggestion
1. Replace the return type of the `deleteBlogPost()` method from `void` to <type://HttpResponse>.
```

ikhoon reviewed (commented):
Awesome work! Left only minor nits.

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-create.mdx:
````diff
@@ -0,0 +1,192 @@
+3. Bind the endpoint `/blogs` to the method.
+
+```java filename=BlogService.java highlight=6
+import com.linecorp.armeria.server.annotation.*;
````
Done for other code samples, too. Had to remove additional import statements in other snippets, e.g. Param, Default and so on. (62784099df81ece9305254ec75ed62bdc35e480f)

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-delete.mdx:
````diff
@@ -0,0 +1,241 @@
+  ```bash
+  $ curl --request POST 'localhost:8080/blogs' \
+  -H 'Content-Type: application/json' \
+  -d '{"title":"Test blog for retrieving", "content":"Create to read."}'
````
Changed the whole example text for READ, UPDATE, and DELETE (62784099df81ece9305254ec75ed62bdc35e480f)

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-delete.mdx:
````diff
@@ -0,0 +1,241 @@
+  ```bash
+  $ curl --request POST 'localhost:8080/blogs' \
+  -H 'Content-Type: application/json' \
+  -d '{"title":"Second blog for test retrieving", "content":"Create to read for the second time"}'
````
Changed the whole example text for READ, UPDATE, and DELETE (62784099df81ece9305254ec75ed62bdc35e480f)

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-delete.mdx:
````diff
@@ -0,0 +1,241 @@
+  For each command, you'll get a response similar to the following.
+
+  ```bash
+  {"id":0,"title":"Test blog for retrieving","content":"Create to read.","createdAt":...,"modifiedAt":...}
````
Changed the whole example text for READ, UPDATE, and DELETE (62784099df81ece9305254ec75ed62bdc35e480f)

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-delete.mdx:
````diff
@@ -0,0 +1,241 @@
+  ```bash
+  {"id":0,"title":"Test blog for retrieving","content":"Create to read.","createdAt":...,"modifiedAt":...}
+
+  {"id":1,"title":"Second blog for test retrieving","content":"Create to read for the second time","createdAt":...,
````
Changed the whole example text for READ, UPDATE, and DELETE (62784099df81ece9305254ec75ed62bdc35e480f)

freevie commented on site/src/pages/tutorials/tutorial-blog-rest-delete.mdx:
````diff
@@ -0,0 +1,241 @@
+  The response contains only one blog post. We can tell that the blog post with the ID `0` has been removed.
+
+  ```bash
+  {"id":1,"title":"Second blog for test retrieving","content":"Create to read for the second time","createdAt":...,"modifiedAt":...}
````
Changed the whole example text for READ, UPDATE, and DELETE (62784099df81ece9305254ec75ed62bdc35e480f)

freevie commented:
> @freevie Could you run `npm run lint:fix` to fix some lint violations?

I hope I did it right. I committed all the changes the linter gave me. (12ae4f43d569cb63143013903aa60bd181c05ee5)

freevie commented:
> A horizontal scroll seems not working.
> <img alt="image" width="800" src="https://user-images.githubusercontent.com/1866157/124559014-30226600-de76-11eb-8ef2-3589382b4827.png">

If you mean the highlighting not covering the whole section, then, yes I am aware of that, and I'd like to leave it as a known issue, if possible?

If you mean the actual scrolling, it works for me. 🧐

ikhoon commented:
> If you mean the actual scrolling, it works for me. 🧐

This is what I wanted to talk about. I used Firefox for testing it.
By the way, I can see the same phenomenon in the master branch.
Let's leave this as a known issue. Let me dig into this later. 😄

ikhoon reviewed (approved):
This is 
![3e6f3cbe00fbcea4d7e72bc150585915](https://user-images.githubusercontent.com/1866157/124696990-0a05d000-df21-11eb-899c-bff9c2d953af.gif)
🚀🚀🚀

ikhoon commented:
@freevie Could you take a look at the following violations?
```
    4:10  error  'RouteComponentProps' is defined but never used                                                                                                                                                              @typescript-eslint/no-unused-vars
  19:9   error  React Hook "useStaticQuery" is called in function "getTutorialSteps" that is neither a React function component nor a custom React Hook function. React component names must start with an uppercase letter  react-hooks/rules-of-hooks
```
https://github.com/line/armeria/pull/3420/checks?check_run_id=3005280774#step:7:5952

You can see the errors by executing `npm run lint`.
https://github.com/line/armeria/blob/b07cc7c9054d70b7debb1e4e1d29d994198bcd3c/site/package.json#L16

trustin commented:
I'm fixing the lint errors. Please wait :-)

trustin commented:
Looking great! I see two issues:
- Could the paths of tutorial pages be `/tutorials/<tutorial_type>/<tutorial_name>/<page_name>`? e.g. `/tutorials/rest/blog/intro`
- How about changing the label 'Tutorial' in the left pane of the documentation section to 'Tutorials'?

trustin commented on site/src/layouts/mdx.tsx:
```diff
@@ -190,9 +190,12 @@ const MdxLayout: React.FC<MdxLayoutProps> = (props) => {
-      mdxNode.href = `/${props.prefix}${
-        mdxNode.parent.name === 'index' ? '' : `/${mdxNode.parent.name}`
-      }`;
+      if (mdxNode.parent.name === 'index') mdxNode.href = `/${props.prefix}`;
+      else if (typeof mdxNode.parent.relativeDirectory === 'undefined')
+        mdxNode.href = `/${props.prefix}/${mdxNode.parent.name}`;
+      else {
```
Let's always use curly braces for `if` and other blocks.
```suggestion
      if (mdxNode.parent.name === 'index') {
        mdxNode.href = `/${props.prefix}`;
      } else if (typeof mdxNode.parent.relativeDirectory === 'undefined') {
        mdxNode.href = `/${props.prefix}/${mdxNode.parent.name}`;
      } else {
```

trustin commented on site/src/pages/tutorials/rest/blog/tutorial-blog-rest-delete.mdx:
```diff
@@ -140,7 +140,7 @@ What if there is no such post to delete? We can check if the blog exists before
     public HttpResponse handleException(ServiceRequestContext ctx, HttpRequest req, Throwable cause) {
       if (cause instanceof IllegalArgumentException) {
-          String message = cause.getMessage();
+        ㄴㅇ  String message = cause.getMessage();
```
```suggestion
          String message = cause.getMessage();
```

trustin commented on site/src/styles/antd-overrides.less:
```diff
@@ -65,6 +65,12 @@
 @marketing-block-width: 532px;
 @marketing-block-margin: 100px;
 
+@import '../components/steps.module.less';
+@process-icon-color: @armeria-magenta;
+@finish-icon-color: @armeria-magenta;
+@finish-tail-color: @armeria-magenta;
+@steps-description-max-width: 105px;
+
```
The color seems too intense to me :sweat_smile: How about using Antd's default colors?
```suggestion
```

trustin commented on site/src/components/steps.module.less:
```diff
@@ -0,0 +1,19 @@
+.ant-steps-horizontal {
```
If you want to customize Antd's Steps component (`@steps-description-max-width: 105px;`), I'd recommend doing it here rather than in the `antd-overrides.less`.

trustin reviewed (commented):
Please feel free to let me know if you want me to update the style of the Steps component.

trustin commented:
@freevie Gentle ping :wink:

trustin commented:
I addressed my comments by myself and fixed the issues with prev/next buttons.

trustin commented:
Thank you so much for your excellent work on the first tutorial. Looking forward to the second one, @freevie, @minwoox and @ikhoon! :heart:
